File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/Object Pascal/Mobile Snippets/ShareSheet/uMain.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit uMain;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants, System.Permissions,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Dialogs, FMX.Objects, FMX.StdCtrls,
  System.Actions, FMX.ActnList, FMX.StdActns, FMX.MediaLibrary.Actions, FMX.Graphics,
  FMX.Controls.Presentation;

type
  TShareSheetForm = class(TForm)
    ActionList1: TActionList;
    ShowShareSheetAction1: TShowShareSheetAction;
    Action1: TAction;
    TakePhotoFromCameraAction1: TTakePhotoFromCameraAction;
    TopToolbar: TToolBar;
    Label1: TLabel;
    BottomToolbar: TToolBar;
    btnShare: TButton;
    btnTakePhoto: TButton;
    imgCameraPicture: TImage;
    procedure ShowShareSheetAction1BeforeExecute(Sender: TObject);
    procedure TakePhotoFromCameraAction1DidFinishTaking(Image: TBitmap);
    procedure btnTakePhotoClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
    FPermissionCamera,
    FPermissionReadExternalStorage,
    FPermissionWriteExternalStorage: string;
    procedure DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
    procedure TakePicturePermissionRequestResult(Sender: TObject; const APermissions: TArray<string>; const AGrantResults: TArray<TPermissionStatus>);
  public
    { Public declarations }
  end;

var
  ShareSheetForm: TShareSheetForm;

implementation

uses
{$IFDEF ANDROID}
  Androidapi.Helpers,
  Androidapi.JNI.JavaTypes,
  Androidapi.JNI.Os,
{$ENDIF}
  FMX.DialogService;

{$R *.fmx}
{$R *.LgXhdpiPh.fmx ANDROID}

procedure TShareSheetForm.FormCreate(Sender: TObject);
begin
{$IFDEF ANDROID}
  FPermissionCamera := JStringToString(TJManifest_permission.JavaClass.CAMERA);
  FPermissionReadExternalStorage := JStringToString(TJManifest_permission.JavaClass.READ_EXTERNAL_STORAGE);
  FPermissionWriteExternalStorage := JStringToString(TJManifest_permission.JavaClass.WRITE_EXTERNAL_STORAGE);
{$ENDIF}
end;

procedure TShareSheetForm.btnTakePhotoClick(Sender: TObject);
begin
  PermissionsService.RequestPermissions([FPermissionCamera, FPermissionReadExternalStorage, FPermissionWriteExternalStorage], TakePicturePermissionRequestResult, DisplayRationale);
end;

procedure TShareSheetForm.ShowShareSheetAction1BeforeExecute(Sender: TObject);
begin
  { show the share sheet }
  ShowShareSheetAction1.Bitmap.Assign(imgCameraPicture.Bitmap);
end;

procedure TShareSheetForm.TakePhotoFromCameraAction1DidFinishTaking(Image: TBitmap);
begin
  { display the picture taken from the camera to the TImage control }
  imgCameraPicture.Bitmap.Assign(Image);
end;

// Optional rationale display routine to display permission requirement rationale to the user
procedure TShareSheetForm.DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
var
  I: Integer;
  RationaleMsg: string;
begin
  for I := 0 to High(APermissions) do
  begin
    if APermissions[I] = FPermissionCamera then
      RationaleMsg := RationaleMsg + 'The app needs to access the camera to take a photo' + SLineBreak + SLineBreak
    else if APermissions[I] = FPermissionReadExternalStorage then
      RationaleMsg := RationaleMsg + 'The app needs to read a photo file from your device';
  end;

  // Show an explanation to the user *asynchronously* - don't block this thread waiting for the user's response!
  // After the user sees the explanation, invoke the post-rationale routine to request the permissions
  TDialogService.ShowMessage(RationaleMsg,
    procedure(const AResult: TModalResult)
    begin
      APostRationaleProc;
    end);
end;

procedure TShareSheetForm.TakePicturePermissionRequestResult(Sender: TObject; const APermissions: TArray<string>; const AGrantResults: TArray<TPermissionStatus>);
begin
  // 3 permissions involved: CAMERA, READ_EXTERNAL_STORAGE and WRITE_EXTERNAL_STORAGE
  if (Length(AGrantResults) = 3) and (AGrantResults[0] = TPermissionStatus.Granted) and (AGrantResults[1] = TPermissionStatus.Granted) and (AGrantResults[2] = TPermissionStatus.Granted) then
    TakePhotoFromCameraAction1.Execute
  else
    TDialogService.ShowMessage('Cannot take a photo because the required permissions are not all granted');
end;

end.

File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/CPP/Mobile Snippets/PhoneDialer/uMain.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <fmx.h>
#ifdef __ANDROID__
#include <Androidapi.Helpers.hpp>
#include <Androidapi.JNI.Os.hpp>
#endif
#include <FMX.DialogService.hpp>
#pragma hdrstop

#include "uMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"

TPhoneDialerForm *PhoneDialerForm;
//---------------------------------------------------------------------------
__fastcall TPhoneDialerForm::TPhoneDialerForm(TComponent *Owner) : TForm(Owner)
{
#ifdef __ANDROID__
    FCallPhonePermission = JStringToString(TJManifest_permission::JavaClass->CALL_PHONE);
#endif
    /* test whether the PhoneDialer services are supported */
    TPlatformServices::Current->SupportsPlatformService(__uuidof(IFMXPhoneDialerService), &FPhoneDialerService);
}
//---------------------------------------------------------------------------
void __fastcall TPhoneDialerForm::DisplayRationale(TObject *Sender, const DynamicArray<String> APermissions, const _di_TProc APostRationaleProc)
{
    // Show an explanation to the user *asynchronously* - don't block this thread waiting for the user's response!
    // After the user sees the explanation, invoke the post-rationale routine to request the permissions
    TDialogService::ShowMessage("The app needs to be able to support your making phone calls",
        [APostRationaleProc](TModalResult AKey)
        {
            APostRationaleProc->Invoke();
        });
}
//---------------------------------------------------------------------------
void __fastcall TPhoneDialerForm::MakePhoneCallPermissionRequestResult(TObject *Sender, const DynamicArray<String> APermissions, const DynamicArray<TPermissionStatus> AGrantResults)
{
	// 1 permission involved: CALL_PHONE
	if ((AGrantResults.Length == 1) && (AGrantResults[0] == TPermissionStatus::Granted))
		FPhoneDialerService->Call(edtTelephoneNumber->Text);
	else
		TDialogService::ShowMessage("Cannot make a phone call because the required permission has not been granted");
}
//---------------------------------------------------------------------------
void __fastcall TPhoneDialerForm::btnGetCarrierInfoClick(TObject *Sender)
{
	/* test whether the PhoneDialer services are supported */
	if (FPhoneDialerService != NULL)
    {
		/* if yes, then update the labels with the retrieved information */
		CarrierNameItem->ItemData->Detail = Format("Carrier Name: %s", ARRAYOFCONST((FPhoneDialerService->GetCarrier()->GetCarrierName())));
		CountryCodeItem->ItemData->Detail = Format("ISO Country Code: %s", ARRAYOFCONST((FPhoneDialerService->GetCarrier()->GetIsoCountryCode())));
		NetworkCodeItem->ItemData->Detail = Format("Network Code: %s", ARRAYOFCONST((FPhoneDialerService->GetCarrier()->GetMobileCountryCode())));
		MobileNetworkItem->ItemData->Detail = Format("Mobile Network: %s", ARRAYOFCONST((FPhoneDialerService->GetCarrier()->GetMobileNetwork())));
	}
	else
    {
        TDialogService::ShowMessage("PhoneDialer service not supported");
    }
}
//---------------------------------------------------------------------------
void __fastcall TPhoneDialerForm::btnMakeCallClick(TObject *Sender)
{
	/* test whether the PhoneDialer services are supported */
	if (FPhoneDialerService != NULL)
    {
		/* if the Telephone Number is entered in the edit box then make the call, else display an error message */
		if (edtTelephoneNumber->Text != "")
        {
            DynamicArray<String> permissions;
            permissions.Length = 1;
            permissions[0] = FCallPhonePermission;

            PermissionsService()->RequestPermissions(permissions, MakePhoneCallPermissionRequestResult, DisplayRationale);
		}
		else
        {
			TDialogService::ShowMessage("Please type-in a telephone number.");
			edtTelephoneNumber->SetFocus();
		}
	}
	else
    {
        TDialogService::ShowMessage("PhoneDialer service not supported");
    }
}
//---------------------------------------------------------------------------
